### ISSUE TITLE
[Proposed Rule] `scss/color-function-syntax`

### ISSUE THREAD

**Issue by lachieh** [stylelint-scss/stylelint-scss#582]
I want to start a discussion before I get into building this rule, since I don't think I've thought through all the potential issues this might add or tangential rules that could solve this problem. If you have some thoughts, please add them. If you're interested in using a rule like this, add a :+1: emoji to the post.

# Background

With the addition of the new CSS Level 4 Color Functions, I initially encountered some confusion around how the Sass versions of these functions work (specifically `rgb`/`rgba` and `hsl`/`hsla`) and what the compiled code ends up being. Using the level 4 spec inside of sass doesn't always do what might be expected.

Sass outputs color values as level 3 syntax when using direct values:
```scss
a { color: rgb(255 255 255 / 0.2); }
// compiles to
a { color: rgba(255, 255, 255, 0.2); }
```

However, when a variable is used it currently interprets the slash character as an operator instead of a divider which means the `blue` and `alpha` variables get interpreted together as an expression and the result is passed into the function as the `blue` value. Using slash as a divider is [deprecated](https://sass-lang.com/documentation/breaking-changes/slash-div) and sass currently throws a warning about this though there is no lint rule currently to warn about this behavior.

```scss
$alpha: 0.2;
a { color: rgb(255 255 255 / $alpha); }
// becomes
a { color: white; }
// this is because 255 / 0.2 = 1275 so sass caps the blue value at 255

```
Additionally, if the value of alpha is anything other than a number then the function will fail because the `blue` value only accepts a [color](https://sass-lang.com/documentation/values/colors), or a number.

```scss
$alpha: 20%;
a { color: rgb(255 255 255 / $alpha); }
// Error: $blue: Expected 12.75%^-1 to have no units or "%".
```

As a side note, it is possible to use interpolation around either the blue or alpha values if the level 4 syntax is desired in the compiled css.

```scss
$alpha: 20%;
a { color: rgb(255 255 255 / #{$alpha}); }
// becomes
a { color: rgb(255 255 255/20%); }
```

It's worth noting that this is incompatible as an [scss color](https://sass-lang.com/documentation/values/colors) so it cannot be nested in another expression or color function without potentially unexpected results.

# Valid Options:
## `modern`
Aligns to Level 4 spec where possible, except where `/` is used as a divider.

Invalid:
```scss
.a { color: rgb(255 255 255 / $alpha); }
 // fixed: .a { color: rgba(255, 255, 255, $alpha); }

.a { color: rgb($red, $green, $blue); }
// fixed: .a { color: rgb($red $green $blue); }
```
Valid:
```scss
.a { color: rgb(255 255 255 / #{$alpha}); }
.a { color: rgb($red, $green, $blue, $alpha); }
.a { color: rgb($color, $alpha); }
```

## `legacy`
Maintains compatibility with `lib-sass` and `ruby-sass`. Only number values accepted as alpha parameter.
From the [sass documentation](https://sass-lang.com/documentation/modules#rgb):
> LibSass and Ruby Sass only support the following signatures:
> 
> `rgb($red, $green, $blue)`
> `rgba($red, $green, $blue, $alpha)`
> `rgba($color, $alpha)`
> Note that for these implementations, the `$alpha` argument is _required_ if the function name rgba() is used, and _forbidden_ if the function name `rgb()` is used.

Invalid:
```scss
.a { color: rgb($red $green $blue); }
// fixed: .a { color: rgb($red, $green, $blue); }

.a { color: rgb($red $green $blue / 20%); }
// fixed: .a { color: .a { color: rgba($red, $green, $blue, 0.2); } }

.a { color: rgba($red, $green, $blue, 20%); }
// fixed: .a { color: .a { color: rgba($red, $green, $blue, 0.2); } }

.a { color: rgba($color, 20%); }
// fixed: .a { color: rgba($color, 0.2); }
```
Valid:
```scss
.a { color: rgb($red, $green, $blue); }
.a { color: rgba($red, $green, $blue, $alpha); }
.a { color: rgba($color, $alpha); }
```

## `percentage`
Same as `legacy`, except that it will accept alpha values specified as percentages.
> Only `dart-sass` and `ruby-sass > 3.7.0` support the alpha value specified as a percentage

Invalid:
```scss
.a { color: rgb(255 255 255); }
// fixed: .a { color: rgb(255, 255, 255); }

.a { color: rgb(255 255 255 / $alpha); }
// fixed: .a { color: .a { color: rgba(255, 255, 255, $alpha); } }
```
Valid:
```scss
.a { color: rgba(255, 255, 255, 20%); }
.a { color: rgba(blue, 20%); }
```

### Extra Details

<details>
<summary>Additional Output Examples</summary>

<table>
<thead>
<tr>
<td>Input</td>
<td>Output</td>
</tr>
</thead>
<tbody>
<tr>
<td>

```scss
$red: 255;
$green: 255;
$blue: 255;
$alpha: 0.2;
$color: white;

/* number values */
.na { color: rgb(255 255 255); }
.nb { color: rgb(255 255 255 / 20%); }
.nc { color: rgb(white, 20%); }
.nd { color: rgb(255, 255, 255); }
.ne { color: rgb(255, 255, 255, 20%); }
.nf { color: rgb(255, 255, 255, $alpha: 20%); }

/* variables */
.va { color: rgb($red $green $blue);  }
.vb { color: rgb($red $green $blue / $alpha); }
.vb2 { color: rgb($red $green $blue / #{$alpha}); }
.vc { color: rgb($color, $alpha); }
.vd { color: rgb($red, $green, $blue); }
.ve { color: rgb($red, $green, $blue, $alpha); }
.vf { color: rgb($red, $green, $blue, $alpha: $alpha); }
```

</td>
<td>

```css

  

  

  

/* number values */
.na { color: white; }
.nb { color: rgba(255, 255, 255, 0.2); }
.nc { color: rgba(255, 255, 255, 0.2); }
.nd { color: white; }
.ne { color: rgba(255, 255, 255, 0.2); }
.nf { color: rgba(255, 255, 255, 0.2); }

/* variables */
.va { color: white; }
.vb { color: white; }
.vb2 { color: rgb(255 255 255/0.2); }
.vc { color: rgba(255, 255, 255, 0.2); }
.vd { color: white; }
.ve { color: rgba(255, 255, 255, 0.2); }
.vf { color: rgba(255, 255, 255, 0.2); }
```

</td>
</tbody>
</table>

</details>

<details>
<summary>Related Issues:</summary>
stylelint-scss/stylelint-config-standard-scss#11
stylelint/stylelint#5671
</details>

**Comment by kristerkari** [comment]
Thanks @lachieh for this very detailed proposal!

I need to properly read through it, but it's good that you have provided a lot of examples of how the rule would work with the different Sass versions.

> ...since I don't think I've thought through all the potential issues this might add or tangential rules that could solve this problem.

This is something that needs to be though of before implementing this kind of rule. It's good to get the discussion started 👍
